left sidebar: Navigation area library rewrite.

Sohaib-Ahmed21 opened the pull request:
Now we don't individually adjust the classes for all the narrow buttons when switching narrows. Multiple calls to function have been replaced by a single call as noted in https://github.com/zulip/zulip/pull/26586#discussion_r1308046455. The **remove** function is also renamed to **deactivate_narrow** function which matches its purpose.

<!-- Describe your pull request here.-->

Fixes: #26686 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
Working in dark mode:

[switching_narrows_dark_mode.webm](https://github.com/zulip/zulip/assets/113179991/719f73a3-22b4-4676-8741-98a07370c68c)

Working in light mode:

[switching_narrrows_light_mode.webm](https://github.com/zulip/zulip/assets/113179991/6a85874b-e474-4430-bbac-734760e2d29e)



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Sohaib-Ahmed21 commented:
@amanagr kindly review this PR.

timabbott commented on web/src/left_sidebar_navigation_area.ts:
```diff
@@ -73,19 +73,21 @@ export function update_dom_with_unread_counts(
+    ];
+    for (const narrow of narrows_to_deactivate) {
+        deactivate_narrow($(narrow));
+    }
```
I'm pretty sure you can do this with jQuery using a single line of code rather than a loop, and that doing so is in fact more performant.

timabbott commented:
Posted one comment on the details, but I think the idea was also to potentially take an argument that's the expected end state view, and not remove that one?

Sohaib-Ahmed21 commented:
@amanagr kindly review this PR.

> Posted one comment on the details, but I think the idea was also to potentially take an argument that's the expected end state view, and not remove that one?

Yeah, that will be more optimistic approach. I'm working on addressing reviews. Thanks!

Sohaib-Ahmed21 commented:
> Posted one comment on the details, but I think the idea was also to potentially take an argument that's the expected end state view, and not remove that one?

Now we take an argument that's the expected end state view, and only remove others except that one.
 
Also I tried using optimistic approach instead of for loop like
```
function deselect_top_left_corner_items(excluded_narrow = ''): void {
    const narrows = [
        ".top_left_all_messages",
        ".top_left_starred_messages",
        ".top_left_mentions",
        ".top_left_recent_view",
        ".top_left_inbox",
    ];


    // Filter out the excluded narrow and join the remaining selectors
    const narrows_to_deactivate = narrows
        .filter((selector) => selector !== excluded_narrow)
        .join(",");

    deactivate_narrow($(narrows_to_deactivate));
}
```
With this code, functionality seems to be correct but test case is failing oddly. Should we proceed with for loop approach as I have done now or can you give some suggestion please?

Sohaib-Ahmed21 commented:
> Posted one comment on the details, but I think the idea was also to potentially take an argument that's the expected end state view, and not remove that one?

I have made the changes as per the review. Kindly review.

alya commented:
Please clean up your commit history and post again to request a review. See [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for guidelines.

Sohaib-Ahmed21 commented:
> Please clean up your commit history and post again to request a review. See [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for guidelines.

I have updated commit messages as per the  guidelines and now commit history has a preparatory commit and a new commit explaining changes properly and I think PR is ready for the review.

alya commented:
Thanks! @amanagr are you up for reviewing this one? (I have not attempted to figure out the context here.)

amanagr commented:
@Sohaib-Ahmed21 how about this:
* Rename `active-filter` to `top-left-active-filter`.
* We don't need to handle `active-sub-filter` here since it not relevant here, so just delete that.
* Do `$(".top-left-active-filter").removeClass("top-left-active-filter")` and then just add the class to the element we want.

amanagr commented:
You should check if this effects active stream element in any way though.

Sohaib-Ahmed21 commented:
> @Sohaib-Ahmed21 how about this:
> 
> * Rename `active-filter` to `top-left-active-filter`.
> * We don't need to handle `active-sub-filter` here since it not relevant here, so just delete that.
> * Do `$(".top-left-active-filter").removeClass("top-left-active-filter")` and then just add the class to the element we want.

Yeah that sounds more logical and hope so will not affect functionality. I will update the PR accordingly while keeping check on active stream element.

Sohaib-Ahmed21 commented:
@amanagr I have addressed the reviews. This PR is ready for review now.

amanagr commented:
> I will update the PR accordingly while keeping check on active stream element.

Looks like you forgot to check for the streams. This does break the active highlight for streams.

Sohaib-Ahmed21 commented:
> > I will update the PR accordingly while keeping check on active stream element.
> 
> Looks like you forgot to check for the streams. This does break the active highlight for streams.

Oh! I forgot to add that in commit. My bad. Will add that as soon as possible.

timabbott commented:
Closing in favor of https://github.com/zulip/zulip/pull/32822, which completed the issue. Thanks for the effort @Sohaib-Ahmed21!
